stratum/hal/lib/tdi/tdi_sde_helpers.cc

```diff
@@ -394,43 +394,65 @@ ::util::Status GetAllEntries(
     std::shared_ptr<::tdi::Session> tdi_session,
     ::tdi::Target tdi_dev_target, const ::tdi::Table* table,
     std::vector<std::unique_ptr<::tdi::TableKey>>* table_keys,
-    std::vector<std::unique_ptr<::tdi::TableData>>* table_values) {
+    std::vector<std::unique_ptr<::tdi::TableData>>* table_datums) {
   CHECK_RETURN_IF_FALSE(table_keys) << "table_keys is null";
-  CHECK_RETURN_IF_FALSE(table_values) << "table_values is null";
+  CHECK_RETURN_IF_FALSE(table_datums) << "table_datums is null";
 
   // Get number of entries. Some types of tables are preallocated and are always
   // "full". The SDE does not support querying the usage on these.
   const ::tdi::Device *device = nullptr;
   ::tdi::DevMgr::getInstance().deviceGet(0, &device);
-  const auto flags = ::tdi::Flags(0);
-  uint32 entries = 0;
+  std::unique_ptr<::tdi::TableKey> table_key;
+  std::unique_ptr<::tdi::TableData> table_data;
+  tdi_status_t tdi_status;
+  std::unique_ptr<::tdi::Target> dev_tgt;
+  device->createTarget(&dev_tgt);
+  uint32 entries = 0, actual = 0;
+  auto table_type = static_cast<tdi_table_type_e>(table->tableInfoGet()->tableTypeGet());
+
+  ::tdi::Flags *flags = new ::tdi::Flags(0);
   if (IsPreallocatedTable(*table)) {
     size_t table_size;
-    RETURN_IF_TDI_ERROR(
-        table->sizeGet(*tdi_session, tdi_dev_target, flags, &table_size));
+    tdi_status =
+        table->sizeGet(*tdi_session, tdi_dev_target, *flags, &table_size);
     entries = table_size;
   } else {
-    RETURN_IF_TDI_ERROR(
-	table->usageGet(*tdi_session, tdi_dev_target, flags, &entries));
+    size_t table_size;
+    tdi_status = table->sizeGet(
+        *tdi_session, tdi_dev_target,
+        *flags, &table_size);
+    entries = table_size;
   }
 
   table_keys->resize(0);
-  table_values->resize(0);
+  table_datums->resize(0);
   if (entries == 0) return ::util::OkStatus();
 
   // Get first entry.
   {
-    std::unique_ptr<::tdi::TableKey> table_key;
-    std::unique_ptr<::tdi::TableData> table_data;
-    RETURN_IF_TDI_ERROR(table->keyAllocate(&table_key));
-    RETURN_IF_TDI_ERROR(table->dataAllocate(&table_data));
-    RETURN_IF_TDI_ERROR(table->entryGetFirst(
-        *tdi_session, tdi_dev_target,
-        flags, table_key.get(),
-        table_data.get()));
+    tdi_status = table->keyAllocate(&table_key);
+    if (tdi_status != TDI_SUCCESS) {
+       goto internal_error;
```

Comment:
Using `goto` is not a good programming model. It is a legacy model used in procedural languages such as C, and not a good fit for C++.

Please revert. You can still use the `RETURN_IF_TDI_ERROR` macros as before. Just change the `::Flags` definition to use const  auto instead.
`const auto flags = ::tdi::Flags(0);`

When you do this, it will automatically get de-allocated after you exit the function. See example here:
https://github.com/ipdk-io/stratum-dev/pull/38/files